File: typescript/problem_40.ts
export function singleNumber(nums: number[]): number {
    let result = 0;
    for (let bit = 0; bit < 32; bit++) {
        const total = nums.reduce((s, n) => s + ((n >> bit) & 1), 0);
        if (total % 3 !== 0) result |= (1 << bit);
    }
    return result;
}

export function problem40(): void {
    console.log("Problem 40:");
    if (singleNumber([6,1,3,3,3,6,6]) !== 1)  throw new Error("assertion failed");
    if (singleNumber([13,19,13,13])   !== 19) throw new Error("assertion failed");
    if (singleNumber([0,0,0,5])       !== 5)  throw new Error("assertion failed");
    console.log("All Tests Passed");
}

File: go/problem_40.go
package main

import "fmt"

// singleNumber finds the element appearing once when all others appear 3 times.
// Uses bit counting: for each bit position, sum all bits mod 3 gives the single number's bit.
func singleNumber(nums []int) int {
	result := 0
	for bit := 0; bit < 64; bit++ {
		sum := 0
		for _, n := range nums { sum += (n >> bit) & 1 }
		if sum%3 != 0 { result |= 1 << bit }
	}
	return result
}

func problem40() {
	fmt.Println("Problem 40:")
	if singleNumber([]int{6, 1, 3, 3, 3, 6, 6}) != 1  { panic("assertion failed") }
	if singleNumber([]int{13, 19, 13, 13}) != 19       { panic("assertion failed") }
	if singleNumber([]int{0, 0, 0, 5}) != 5            { panic("assertion failed") }
	fmt.Println("All Tests Passed")
}
